File: NetJS/API/IO.cs
using System;
using Microsoft.ClearScript.JavaScript;

namespace NetJS.API
{
    /// <summary>IO class contains methods for file and directory manipulation.</summary>
    /// <example>IO can read, write and delete a file:
    /// <code lang="javascript">var file = "example.txt";
    /// IO.write(file, "Hello World!");
    /// console.log(IO.read(file); //prints: Hello World!
    /// IO.deleteFile(file);</code></example>
    public class IO {

        private static string GetFile(string name) {
            return State.Application.Settings.Root + name;
        }

        /// <summary>Writes text into a file.</summary>
        /// <param name = "file">A filename (string)</param>
        /// <param name = "content">The text to be written (string)</param>
        /// <example><code lang="javascript">IO.writeText("data.json", "Hello World!");</code></example>
        public static dynamic writeText(string name, string content) {
            var application = State.Application;
            var state = State.Get();
            var file = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    System.IO.File.WriteAllText(file, content);
                    application.AddCallback(resolve, true, state);
                } catch {
                    application.AddCallback(reject, $"Could not write text to file '{file}'", state);
                }
            });
        }

        /// <summary>Writes bytes into a file.</summary>
        /// <param name = "file">A filename</param>
        /// <param name = "content">The bytes to be written (Uint8Array)</param>
        /// <example><code lang="javascript">IO.writeBytes("image.png", bytes);</code></example>
        public static dynamic writeBytes(string name, dynamic bytes) {
            var application = State.Application;
            var state = State.Get();
            var file = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                var content = (IArrayBuffer)bytes.buffer;

                try {
                    System.IO.File.WriteAllBytes(file, content.GetBytes());
                    application.AddCallback(resolve, true, state);
                } catch {
                    application.AddCallback(reject, $"Could not write bytes to file '{file}'", state);
                }
            });
        }

        /// <summary>Reads and returns text content of a file.</summary>
        /// <param name= "file">A filename to read from (string)</param>
        /// <returns>The content of the file (string)</returns>
        /// <example><code lang="javascript">var text = IO.readText("data.json");</code></example>
        public static dynamic readText(string name) {
            var application = State.Application;
            var state = State.Get();
            var file = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    application.AddCallback(resolve, System.IO.File.ReadAllText(file), state);
                } catch {
                    application.AddCallback(reject, $"Could not read text from file '{file}'", state);
                }
            });
        }

        /// <summary>Reads and returns binary content of a file.</summary>
        /// <param name= "file">A filename to read from (string)</param>
        /// <returns>The binary content of the file (Uint8Array)</returns>
        /// <example><code lang="javascript">var bytes = IO.readBytes("image.png");</code></example>
        public static dynamic readBytes(string name) {
            var application = State.Application;
            var state = State.Get();
            var file = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    var bytes = System.IO.File.ReadAllBytes(file);
                    application.AddCallback(resolve, Tool.ToByteArray(bytes), state);
                } catch (Exception e) {
                    application.AddCallback(reject, $"Could not read bytes from file '{file}'", state);
                }
            });
        }

        /// <summary>Deletes a file.</summary>
        /// <param name= "file">A filename to delete</param>
        /// <example><code lang="javascript">IO.deleteFile("data.json");</code></example>
        public static dynamic deleteFile(string name) {
            var application = State.Application;
            var state = State.Get();
            var file = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    System.IO.File.Delete(file);
                    application.AddCallback(resolve, true, state);
                } catch {
                    application.AddCallback(reject, $"Could not delete file '{file}'", state);
                }
            });
        }

        /// <summary>Copies a file.</summary>
        /// <param name= "source">The file to copy</param>
        /// <param name= "destination">The file to copy to</param>
        /// <example><code lang="javascript">IO.copyFile("a.txt", "b.txt");</code></example>
        public static dynamic copyFile(string source, string destination) {
            var application = State.Application;
            var state = State.Get();
            var fileA = GetFile(source);
            var fileB = GetFile(destination);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    System.IO.File.Copy(fileA, fileB);
                    application.AddCallback(resolve, true, state);
                } catch {
                    application.AddCallback(reject, $"Could not copy '{fileA}' to '{fileB}'", state);
                }
            });
        }

        /// <summary>Moves/renames a file.</summary>
        /// <param name= "source">The source location</param>
        /// <param name= "destination">The destination</param>
        /// <example><code lang="javascript">IO.moveFile("a.txt", "files/b.txt");</code></example>
        public static dynamic moveFile(string source, string destination) {
            var application = State.Application;
            var state = State.Get();
            var fileA = GetFile(source);
            var fileB = GetFile(destination);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    System.IO.File.Move(fileA, fileB);
                    application.AddCallback(resolve, true, state);
                } catch {
                    application.AddCallback(reject, $"Could not move '{fileA}' to '{fileB}'", state);
                }
            });
        }

        /// <summary>Moves/renames a directory.</summary>
        /// <param name= "source">The source location</param>
        /// <param name= "destination">The destination</param>
        /// <example><code lang="javascript">IO.moveDirectory("files", "documents/files");</code></example>
        public static dynamic moveDirectory(string source, string destination) {
            var application = State.Application;
            var state = State.Get();
            var dirA = GetFile(source);
            var dirB = GetFile(destination);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    System.IO.Directory.Move(dirA, dirB);
                    application.AddCallback(resolve, true, state);
                } catch {
                    application.AddCallback(reject, $"Could not move directory '{dirA}' to '{dirB}'", state);
                }
            });
        }

        /// <summary>Get all files in a directory.</summary>
        /// <param name= "directory">The directory path</param>
        /// <example><code lang="javascript">var files = IO.getFiles("documents");</code></example>
        public static dynamic getFiles(string directory) {
            var application = State.Application;
            var state = State.Get();
            var dir = GetFile(directory);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    application.AddCallback(resolve, Tool.ToArray(System.IO.Directory.GetFiles(dir)), state);
                } catch {
                    application.AddCallback(reject, $"Could not get files from directory '{dir}'", state);
                }
            });
        }

        /// <summary>Get all directories in a directory.</summary>
        /// <param name= "directory">The directory path</param>
        /// <example><code lang="javascript">var directories = IO.getDirectories("documents");</code></example>
        public static dynamic getDirectories(string directory) {
            var application = State.Application;
            var state = State.Get();
            var dir = GetFile(directory);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    application.AddCallback(resolve, Tool.ToArray(System.IO.Directory.GetDirectories(dir)), state);
                } catch {
                    application.AddCallback(reject, $"Could not get directories from directory '{dir}'", state);
                }
            });
        }

        /// <summary>Checks if the file exists.</summary>
        /// <param name= "file">The file path</param>
        /// <example><code lang="javascript">var exists = IO.fileExists("name.txt");</code></example>
        public static dynamic fileExists(string name) {
            var application = State.Application;
            var state = State.Get();
            var file = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    application.AddCallback(resolve, System.IO.File.Exists(file), state);
                } catch {
                    application.AddCallback(reject, $"Could not check if file '{file}' exists", state);
                }
            });
        }

        /// <summary>Checks if the directory exists.</summary>
        /// <param name= "directory">The directory path</param>
        /// <example><code lang="javascript">var exists = IO.directoryExists("documents");</code></example>
        public static dynamic directoryExists(string name) {
            var application = State.Application;
            var state = State.Get();
            var dir = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    application.AddCallback(resolve, System.IO.Directory.Exists(dir), state);
                } catch {
                    application.AddCallback(reject, $"Could not check if directory '{dir}' exists", state);
                }
            });
        }

        /// <summary>Creates a new directory.</summary>
        /// <param name= "directory">The directory path</param>
        /// <example><code lang="javascript">IO.createDirectory("documents");</code></example>
        public static dynamic createDirectory(string name) {
            var application = State.Application;
            var state = State.Get();
            var dir = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    System.IO.Directory.CreateDirectory(dir);
                    application.AddCallback(resolve, true, state);
                } catch {
                    application.AddCallback(reject, $"Could not create directory '{dir}'", state);
                }
            });
        }

        /// <summary>Deletes a directory.</summary>
        /// <param name= "directory">The directory path</param>
        /// <example><code lang="javascript">IO.deleteDirectory("documents");</code></example>
        public static dynamic deleteDirectory(string name) {
            var application = State.Application;
            var state = State.Get();
            var dir = GetFile(name);

            return Tool.CreatePromise((resolve, reject) => {
                try {
                    System.IO.Directory.Delete(dir);
                    application.AddCallback(resolve, true, state);
                } catch {
                    application.AddCallback(reject, $"Could not delete directory '{dir}'", state);
                }
            });
        }
    }
}

File: Testing/test/src/io.js
async function testIO(){
	// write + read text
	let writeTextResult = await IO.writeText("io/a.txt", "testvalue38");
	Test.assert(() => writeTextResult === true, "IO.writeText");
	let readTextResult = await IO.readText("io/a.txt");
	Test.assert(() => readTextResult === "testvalue38", "IO.readText");

	// write + read bytes
	let bytes = new Uint8Array(1);
	bytes[0] = 42;
	let writeBytesResult = await IO.writeBytes("io/bytes.txt", bytes);
	Test.assert(() => writeBytesResult === true, "IO.writeBytes");
	let readBytesResult = await IO.readBytes("io/bytes.txt");
	Test.assert(() => readBytesResult[0] === 42, "IO.readBytes");

	// copy file
	let copyFileResult = await IO.copyFile("io/a.txt", "io/b.txt");
	Test.assert(() => copyFileResult === true, "IO.copy");
	let copiedFileText = await IO.readText("io/b.txt");
	Test.assert(() => copiedFileText === "testvalue38", "IO.copy new file");

	// delete file
	let deleteFileResult = await IO.deleteFile("io/a.txt");
	Test.assert(() => deleteFileResult === true, "IO.delete");
	let deletedFileGone = await IO.fileExists("io/a.txt");
	Test.assert(() => deletedFileGone === false, "IO.delete gone");

	// move file
	let moveFileResult = await IO.moveFile("io/b.txt", "io/c.txt");
	Test.assert(() => moveFileResult === true, "IO.move");
	let movedFileGone = await IO.fileExists("io/b.txt");
	Test.assert(() => movedFileGone === false, "IO.move old gone");
	let movedFileText = await IO.readText("io/c.txt");
	Test.assert(() => movedFileText === "testvalue38", "IO.move new file");

	await IO.deleteFile("io/c.txt");

	// create directory
	let createDirectoryResult = await IO.createDirectory("io/documents");
	Test.assert(() => createDirectoryResult === true, "IO.createDirectory");
	let createdDirectoryExists = await IO.directoryExists("io/documents");
	Test.assert(() => createdDirectoryExists === true, "IO.createDirectory exists");

	// move directory
	let moveDirectoryResult = await IO.moveDirectory("io/documents", "io/pictures");
	Test.assert(() => moveDirectoryResult === true, "IO.moveDirectory");
	let movedDirectoryGone = await IO.directoryExists("io/documents");
	Test.assert(() => movedDirectoryGone === false, "IO.moveDirectory old gone");
	let movedDirectoryExists = await IO.directoryExists("io/pictures");
	Test.assert(() => movedDirectoryExists === true, "IO.moveDirectory new exists");

	// delete directory
	let deleteDirectoryResult = await IO.deleteDirectory("io/pictures");
	Test.assert(() => deleteDirectoryResult === true, "IO.deleteDirectory");
	let deletedDirectoryExists = await IO.directoryExists("io/pictures");
	Test.assert(() => deletedDirectoryExists === false, "IO.deleteDirectory gone");

	// get files
	let files = await IO.getFiles("io/example");
	Test.assert(() => files.length == 2, "IO.getFiles count");
	Test.assert(() => files[0].endsWith("fileA.txt"), "IO.getFiles[0]");
	Test.assert(() => files[1].endsWith("fileB.png"), "IO.getFiles[1]");

	// get directories
	let directories = await IO.getDirectories("io/example");
	Test.assert(() => directories.length == 2, "IO.getDirectories count");
	Test.assert(() => directories[0].endsWith("folderA"), "IO.getDirectories[0]");
	Test.assert(() => directories[1].endsWith("folderB"), "IO.getDirectories[1]");
}